### ISSUE TITLE
Types

### ISSUE THREAD

**Issue by BookOwl** [haumea-lang/haumea#4]
Haumea is pretty limited with just integers right now. I plan on adding Floats, Strings, and Arrays.

**Comment by nanalan** [comment]
Strings could be Arrays of characters?

Float/integer division, too (`9 / 2 = 4` vs `9 // 2 = 4.5`).

**Comment by joker314** [comment]
Othere way around @nanalan I think?

**Comment by nanalan** [comment]
@joker314 quite possibly.

**Comment by BookOwl** [comment]
OK, here is a some syntax that I've come up with for declaring types:
```
/* Variable declarations
variable x is an Integer /* Same as "long x" in c */

variable foo is a String /* Same as "char* foo" in C */

variable y /* Same as we have now, makes y an Integer */

/* Function parameters */
to foo with (name is a String, foo is a Float, bar is an Integer, x) returns a Float ...
```
I'm not very happy with the function parameter syntax, do you guys have any better ideas?

**Comment by TheMonsterFromTheDeep** [comment]
Maybe 'as'?
```
to foo as Float with (name as String, foo as Float, bar as Integer) ...
```
this could also work for normal variables too:
```
variable x as Integer
```

**Comment by Jonathan50** [comment]
Why static typing?

**Comment by jupiterio** [comment]
@Jonathan50 I'm guessing it's because it compiles directly to C. Might be a bit dumb to try and make it dynamically typed

**Comment by Jonathan50** [comment]
> Might be a bit dumb to try and make it dynamically typed

Huh?

**Comment by TheMonsterFromTheDeep** [comment]
I think what IO means is that it's kind of a mess to have dynamic typing in a compiled language, even if it's just compiled to C.

**Comment by jupiterio** [comment]
Specially if the language that it's compiled to is statically typed.

**Comment by BookOwl** [comment]
> Why static typing?

Because Haumea is a compiled language, and it's kind of hard to have dynamic typing in a compiled language (especially when the target is C)

I guess I could try and add type inference.

**Comment by Jonathan50** [comment]
Okay. (You could just have a struct with a union member and a tag member) :+1:

**Comment by nanalan** [comment]
Type inference please, too. It'd be nice to have a more consise syntax for defining variables as its a common thing..

**Comment by nanalan** [comment]
Regarding the syntax for types, why not drop the **variable** tag altogether? It still reads correctly:

```ruby
/* Variable declarations
integer x /* Same as "long x" in c */

string foo /* Same as "char* foo" in C */

/* Function parameters */
to foo with (string name, float foo, integer bar) as string...
```

**Comment by BookOwl** [comment]
That looks great, but I still don't like the syntax for the return type for functions.

**Comment by nanalan** [comment]
Agreed. If we get inference working there's no need for it anyway

**Comment by nanalan** [comment]
Is it worth having more simplistic types? e.g. _string_, _list_, and _number_?

**Comment by BookOwl** [comment]
Yes for string and list. I'm not so sure about combining floats and integers because integers can be much more accurate.

**Comment by nanalan** [comment]
Nevermind that, seperate number types would (and still are for some) pretty confusing - such as two different kinds of division etc..

**Comment by TheMonsterFromTheDeep** [comment]
So are you guys going to use a pre-existing C library for these types, or write your own from the ground up?

(by the way, list libraries are a big mess because there's no templating system...)

**Comment by BookOwl** [comment]
For strings, ints, and floats I'm planning on using the C types. I'm not sure about for lists.

**Comment by TheMonsterFromTheDeep** [comment]
> @TheMonsterFromTheDeep, what do you think would be best? You have much more experience with C than I do.

There are a few different ways:

First, there is a raw macro-based approach like the guy is using in [this stackoverflow post](http://stackoverflow.com/questions/27890462/using-macros-to-implement-a-generic-vector-in-c-is-this-a-good-idea). This one is probably the fastest at run-time, but it could be a bad idea because it bloats compile time - whenever a list 'function' is used, it has to be re-compiled because it's actually a macro.

One way I've thought of before is using macros in a template-like system - essentially, there are a set of macros to generate list methods for some individual type, and then whenever a list of that type is used, those individual functions are used. This, I think, would also be pretty fast at runtime, and would reduce the number of functions compiled from one per use to like twenty per type.

Finally, it's also possible to have list types that are completely generic - they just use a void pointer and whenever the list is used it has it's pointer cast to the correct pointer type. I'm honestly not sure what the tradeoff is here - I would guess it would technically be slower at runtime because it has to cast the pointer every time, but I'm not sure if the performance hit would actually be noticeable.

I suppose it would be possible to put together a bunch of different tests for each of these different methods and see how they compare.

**Comment by Jonathan50** [comment]
> but it could be a bad idea because it bloats compile time - whenever a list 'function' is used, it has to be re-compiled because it's actually a macro.

What about having a `define_list_of` macro that defines a new global type and all its operations, and have the code generator scan the annotated program for used types and define all needed types only once?

**Comment by TheMonsterFromTheDeep** [comment]
> > but it could be a bad idea because it bloats compile time - whenever a list 'function' is used, it has to be re-compiled because it's actually a macro.
> 
> What about having a define_list_of macro that defines a new global type and all its operations, and have the code generator scan the annotated program for used types and define all needed types only once? (And replace calls to list operations with calls to the appropriate C procedures depending on the types of the arguments.)

Yeah, that's what my second method meant:
> One way I've thought of before is using macros in a template-like system - essentially, there are a set of macros to generate list methods for some individual type, and then whenever a list of that type is used, those individual functions are used. This, I think, would also be pretty fast at runtime, and would reduce the number of functions compiled from one per use to like twenty per type.

Maybe I wasn't being very clear. :P

**Comment by Jonathan50** [comment]
Oh, sorry. I didn't read your whole post.

**Comment by TheMonsterFromTheDeep** [comment]
> Casting just tells the C compiler what type an expression's value is.

Oh. Derp. :P

However, in that third method, the size would have to be passed to each function as a parameter, so I think it would still be slightly slower. ;)

**Comment by BookOwl** [comment]
How about this for function parameter type syntax?
```
to foo with (name is a String, foo is a Float, bar is an Integer, return is a Float) ...

**Comment by BookOwl** [comment]
Moved to https://github.com/haumea-lang/thinktank/issues/4
